x/metadata/keeper/signers_utils.go

```diff
@@ -0,0 +1,348 @@
+	// If an equal party is already in the details, just update its optional flag
+	// to false, otherwise, add it to the list.
+reqPartiesLoop:
+	for _, reqParty := range reqParties {
+		if !reqParty.Optional {
+			for _, party := range details {
+				if reqParty.IsSameAs(party) {
+					party.MakeRequired()
+					continue reqPartiesLoop
+				}
+			}
+			details = append(details, WrapRequiredParty(reqParty))
+		}
+	}
+
+	return details
+}
+
+func (p *PartyDetails) SetAddress(address string) {
+	if p.address != address {
+		p.acc = nil
+	}
+	p.address = address
+}
+
+func (p *PartyDetails) GetAddress() string {
+	if len(p.address) == 0 && len(p.acc) > 0 {
+		p.address = p.acc.String()
+	}
+	return p.address
+}
+
+func (p *PartyDetails) SetAcc(addr sdk.AccAddress) {
+	if !bytes.Equal(p.acc, addr) {
+		p.address = ""
+	}
+	p.acc = addr
+}
+
+func (p *PartyDetails) GetAcc() sdk.AccAddress {
+	if len(p.acc) == 0 && len(p.address) > 0 {
+		p.acc, _ = sdk.AccAddressFromBech32(p.address)
+	}
+	return p.acc
+}
+
+func (p *PartyDetails) SetRole(role types.PartyType) {
+	p.role = role
+}
+
+func (p *PartyDetails) GetRole() types.PartyType {
+	return p.role
+}
+
+func (p *PartyDetails) SetOptional(optional bool) {
+	p.optional = optional
+}
+
+func (p *PartyDetails) MakeRequired() {
+	p.optional = false
+}
+
+func (p *PartyDetails) GetOptional() bool {
+	return p.optional
+}
+
+func (p *PartyDetails) IsRequired() bool {
+	return !p.optional
+}
+
+func (p *PartyDetails) SetSigner(signer string) {
+	if p.signer != signer {
+		p.signerAcc = nil
+	}
+	p.signer = signer
+}
+
+func (p *PartyDetails) GetSigner() string {
+	if len(p.signer) == 0 && len(p.signerAcc) > 0 {
+		p.signer = p.signerAcc.String()
+	}
+	return p.signer
+}
+
+func (p *PartyDetails) SetSignerAcc(signerAddr sdk.AccAddress) {
+	if !bytes.Equal(p.signerAcc, signerAddr) {
+		p.signer = ""
+	}
+	p.signerAcc = signerAddr
+}
+
+func (p *PartyDetails) GetSignerAcc() sdk.AccAddress {
+	if len(p.signerAcc) == 0 && len(p.signer) > 0 {
+		p.signerAcc, _ = sdk.AccAddressFromBech32(p.signer)
+	}
+	return p.signerAcc
+}
+
+func (p *PartyDetails) HasSigner() bool {
+	return len(p.signer) > 0 || len(p.signerAcc) > 0
+}
+
+func (p *PartyDetails) CanBeUsed() bool {
+	return p.canBeUsedBySpec
+}
+
+func (p *PartyDetails) MarkAsUsed() {
+	p.usedBySpec = true
+}
+
+func (p *PartyDetails) IsUsed() bool {
+	return p.usedBySpec
+}
+
+// IsStillUsableAs returns true if this party can be used, hasn't yet been used and has the provided role.
+func (p *PartyDetails) IsStillUsableAs(role types.PartyType) bool {
+	return p.CanBeUsed() && !p.IsUsed() && p.GetRole() == role
+}
+
+// IsSameAs returns true if this is the same as the provided Party or PartyDetails.
+// Only the address and role are considered for this test.
+func (p *PartyDetails) IsSameAs(p2 types.Partier) bool {
+	return types.SamePartiers(p, p2)
+}
+
+// SignersWrapper is a thing that holds the signers as strings and acc addresses.
```

Comment:
Maybe rephrase this as "SignersWrapper stores the signers as strings and acc addresses."